### PR TITLE
[WIP] Cortical Borers

### PR DESCRIPTION
**Pull request by oisin100**
No-one else could arsed doing this :(

I've started work on the cortical borers but i need some salt from you all to make it actually work out! First things first goons toxins are different then ours so what toxins should borers be able to secrete? Current ones: `"mannitol","bicaridine","hyperzine", "kelotane", "morphine", "cyanide"`
### Cool Checklist for borers
- [x] Allow borers to infect people
- [x] Allow borers to secrete reagents into people
- [x] Allow borers to "hide"
- [x] Update borers sprite from changeling slug to the actual sprite
- [x] Allow the borers to jumpstart(Revive) their host with 250 chems 
- [x] Allow borers to temporarily take over their host
- [x] Allow borers to reproduce(This could be done by either gibbing the host or maybe making them vomit up a new borer...)
- [x] Stop borers from talking outside of host.
- [x] Listen to adams salt and nerf borers so bad that no-one will ever want to play as one.

Note: Code is in a hacky unrefined state at the moment... Ill make the code more efficient soon..

### CODE REVIEW

**Comment by Time-Green** [comment]
This is so beautifull. I am not a coder.

But you got my moral support

**Comment by ShadowDeath6** [comment]
100/10 I thought this died in the issue request. Coders are doing god's work.

Probably replace cyanide with nulvoscine or whatever reagent sleepy pens inject, the borer isn't supposed to kill their host, realistically.

**Comment by Cruix** [comment]
Hiding is pretty easy, check out https://github.com/yogstation13/yogstation/blob/a19f05dc74b2a70b4565bfdfc978f996de5acf78/code/modules/mob/living/carbon/alien/larva/powers.dm

On the topic of balance, rather than always being made docile by sugar, perhaps it should be a random reagent from a list to discourage metagaming. To take it one step further, add another random reagent from the same list that gives the borer free chemicals.

"No, don't eat all that sugar and then jump out an airlock, I'll...psyche, thanks for the free chems."

**Comment by ShadowDeath6** [comment]
Borers actually don't give a shit about suicide since when they revive the body they're pretty much space immune.

**Comment by oisin100** [comment]
Only carbon animals can use powers like that @Cruix. So ill just copy paste the code...

The extra reagent sounds like a good idea...

Borers really wouldn't give a shit about their master committing sucide as they are(Or will be...) Space resistant. Sugar only makes them unable to use their abilities.

**Comment by ShadowDeath6** [comment]
Definitely have the host just vomit up a borer, punishing host with gibbing would be harsh.

**Comment by oisin100** [comment]
Yea i was thinking that too. Would also stop people from spacing themselves because "lol borer"

**Comment by ShadowDeath6** [comment]
Will revival just put the borer in control or will it go the full 100%, like giving them a constant healing / space immunity thing? It can work fine either way.
On bay, reviving the host puts the borer in permanent control, giving them a constant amount of bicaridine, antitoxin, inaprovaline (epinephrine here), and antitoxin (which is the borg kind, not charcoal). It also makes them leak blood and when you look at them it says "they appear to be leaking some sort of fluid".

**Comment by oisin100** [comment]
It will heal them up to 100% if the borer wants to escape space they would just have to leave their current host.

**Comment by ShadowDeath6** [comment]
Also, maybe add stimulant so it can let the host move faster? Just a thought.

Edit- This would probably be replacing hyperzine since they do a super similar thing and hyperzine isn't obtainable in goonchem.

**Comment by oisin100** [comment]
Sounds like a good idea!

**Comment by ShadowDeath6** [comment]
Also- Make sure that borers can't bite or talk outside of a host. Not sure how you'd go about doing that, but... keep it in mind.

**Comment by ghost** [comment]
so if you get infected by a borer there's no real way to stop it and if you suicide it gets to get back up free of charge no cooperation necessary and space immune and healing on it's own?

why is this good.

seriously, why is this good at all.

what would it's objectives be.

what incentive do people have to not just cremate themselves or throw themselves into a gibber or wall themselves in.

and if your answer to 'well what stops them screaming over comms' is 'it gets to inject mute toxin lol'' it might as well be the borer just ghosts them until they leave the body because what's the point of giving the borer all these cooperation abilities and chems ect if it's just denying anything at all to the host

tl;dr this is really shitty and again not going to work out on a server like yog because again we're lowrp unlike bay or paradise and this is an rp dependant antag, and if you make it non-rp dependant it'll just suck.

fully and entirely against this showing up, as i always have been, because borers just look dumb even on paradise and some asshole can just wall themselves in out of spite, and if your response is 'lol mental domination' just make it fucking ghost the guy in the body if you can't do anything the borer doesn't want you to anyway

**Comment by oisin100** [comment]
> so if you get infected by a borer there's no real way to stop it and if you suicide it gets to get back up free of charge no cooperation necessary and space immune and healing on it's own?

No, When it revives the host it wont take control unless theres no mind in the host.

> what incentive do people have to not just cremate themselves or throw themselves into a gibber or wall themselves in.

The borer can do literally nothing to harm you except taking control of your body every now and then for a couple of seconds and making you sleep. But if you establish a good relationship with your borer it can provide some great buffs such as self-healing, stimulants and revival(This should probably be restricted to one use). Perhaps the borer wouldn't be able to infect people with loyalty implants and have sec treat the borer as a hostile threat.

> and if your answer to 'well what stops them screaming over comms' is 'it gets to inject mute toxin lol'' it might as well be the borer just ghosts them until they leave the body because what's the point of giving the borer all these cooperation abilities and chems ect if it's just denying anything at all to the host

Borers dont have mute toxins. Sure they have morphine but that wont shut them up straight away.

> fully and entirely against this showing up, as i always have been, because borers just look dumb even on paradise and some asshole can just wall themselves in out of spite, and if your response is 'lol mental domination' just make it fucking ghost the guy in the body if you can't do anything the borer doesn't want you to anyway

It might take a bit but you can always just walk to medbay and have the borer surgically removed.

**Comment by oisin100** [comment]
Probably a good idea to make the borers not regenerate chemicals whilst the host is unconscious

**Comment by ShadowDeath6** [comment]
If you cremate / gib yourself who really gives a shit? The borer is dead, so are you. Whoop de doo. You don't need to cooperate with a borer, it just makes it easier. They're not supposed to be vg's happy little helper, they're supposed to be slugs that control people. This isn't hard to understand. 

As per making it so that borers don't just lelkill the host and assume direct control, make the reanimated corpse have some fancy flavortext so that they get lynched, or make it so that borers can't revive unless they have 250 chems (the max amount available), making it so they have some incentive to stay inside the person and actually help.

**Comment by Xantam-13** [comment]
What's a borer?

**Comment by Time-Green** [comment]
http://ss13.moe/wiki/index.php/Cortical_Borer
This is sorta it.

**Comment by ghost** [comment]
Replace morphine with tirezine so the borer can legit sedate the host without getting KO'd.

Give it pleufloradecalin for a muting reagent to prevent host screaming over comms.

Give it leporazine for spacewalking as a reagent.

Make hijack last as long as the host doesn't resist-basically like cuffs, unless you resist you're stuck in them forever. So if the borer hijacks, it stays in control until the host resists, which takes a minute to break out of mental control.

**Comment by ShadowDeath6** [comment]
Perflouro is muting only, nulvoscine or whatever it's called will keep the host from *me'ing as well. 

Also the abilities in this PR are more congruous with http://nanotrasen.se/wiki/index.php/Cortical_Borer

Also tirizine is literally the most op KO chemical, that's the shit in sleep sting.

**Comment by oisin100** [comment]
Introduced a new "influence" system. Borers can now talk telepathically to their hosts. Talking telepathicaly can be whispering to shouting depending on the borers influence to the host.
Some chems require the borer to have more than a certain ammount of influence such as morphine what required you to have 50% influence.
Using certain chemicals will affect your influence with your host. For example using healing chems will increase your influence with your host by 5% whilst using morphine will decrease it by 10%
My plan is to have the borer only able to assume control when it has 100% influence.

File: code/modules/mob/living/simple_animal/borer/borer_powers.dm [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+	if(docile)
+		src << "<span class='boldnotice'>You are feeling far too docile to do that.</span>"
+		return
+
+	if(chemicals < 50)
+		src << "<span class='boldnotice'>You don't have enough chemicals!</span>"
+		return
+
+	var/list/chemnames = list()
+	for(var/datum in typesof(/datum/borer_chem))
```
Comment (oisin100) [inline code/modules/mob/living/simple_animal/borer/borer_powers.dm:70]: Ill update this code to be more efficient at some point... It might cause memory leaks..

**Comment by ShadowDeath6** [comment]
That could actually work out really well. Will assuming control cost influence?
Also, I recommend you drop the influence req to like 75-90 since it lets the borer do some things without jeopardizing its chances of assuming control.

**Comment by oisin100** [comment]
Alright, Im not sure about assuming control costing influence... I suppose it should.

File: code/modules/mob/living/simple_animal/borer/borer_powers.dm (hunk 1 above)
Comment (Cruix) [inline code/modules/mob/living/simple_animal/borer/borer_powers.dm:70]: Byond has a built-in garbage collector, so you don't have to worry about memory leaks (though we avoid using it as much as possible, because qdel() is better). As far as I can tell, this shouldn't cause any problems. There are many places in the code that handle input() lists similarly. If you are going to change it, I would recommend making an associative list of chem names to chem datums instead of using typesof() twice.

Comment (oisin100) [inline code/modules/mob/living/simple_animal/borer/borer_powers.dm:70]: Good to know! Yea, I can see why having a dedicate list for it would be more efficient!

**Comment by ghost** [comment]
that's bad no

use paradise's borer chem system rather than influence because otherwise the borer can't do shit

**Comment by ShadowDeath6** [comment]
I think he's using influence so the borer has to cooperate if it wants to breed/do anything fun for itself.

**Comment by ShadowDeath6** [comment]
As for the chem list, potassium iodide isn't really needed due to it being a super niche chem. I'd still say go with meth or ephedrine for dank speed boosts.

**Comment by ghost** [comment]
Stimulants are a more potent speedboost/stunresist chem that has a lower OD threshold but less terrible for being used effects than meth.

**Comment by oisin100** [comment]
Potassium iodide is there for the borer to repair radiation.
As for stimulants i think we should go for Ephedrine as this is what the stimulant virus gives.
Also when a cortical borer event starts how should the corer be spawned? In a random vent on the station or already in someones head?

**Comment by ShadowDeath6** [comment]
I'd say in some place with vents, like oldbar or something. Definitely not in somebody's head, that would make no sense.

Also, radiation really doesn't matter that much since it just turns into toxins and the borer can just repair toxins damage with charcoal.

**Comment by Time-Green** [comment]
What about chefs backroom?
If I was a parasite I would hatch there.

While im saying it, could it be a thing where you hide in food?

**Comment by oisin100** [comment]
Alright, ill remove potas iodide.

So your saying that it should spawn in a couple of predefine spots then? Perhaps these spots could be:
- Kitchen Backroom
- Morgue
- Old bar
- Some places in the maints.

Perhaps we could make it so they can hide in food... But im thinking that will need a bit of effort to do. And coders dont like effort.

**Comment by Time-Green** [comment]
The hiding in food thing was just an idea.
Cant see why a borer would take the effort to do that anyway.
But yeah, the predefined spots need to be gross or weird places.
Perhaps  janitorial closet and the toilets aswell?

Or vacant office, because its just lonely and ez to start?

**Comment by ShadowDeath6** [comment]
Hype

You probably don't REALLY need borer objectives since they're basically a swarmer-tier "antag" mob that just exists.

**Comment by oisin100** [comment]
Suppose so. Its probably not a good idea to give them objectives anyway. Theres enough spam at round end as is.

Just need to fix the merge conflicts then we should be done(Apart from testing)

File: code/modules/mob/living/simple_animal/borer/borer.dm [hunk 2]
```diff
@@ -0,0 +1,203 @@
+/mob/living/simple_animal/borer
+	name = "Cortical Borer"
+	desc = "Absolutely not de-beaked or harmless. Keep away from corpses."
```
Comment (ShadowDeath6) [inline code/modules/mob/living/simple_animal/borer/borer.dm]: Might want to edit the description so it's not the same as ling headslug

On other servers it's "A small, quivering, slug-like creature."

Comment (oisin100) [inline code/modules/mob/living/simple_animal/borer/borer.dm]: Strange... i thought i did... Will do!

File: code/modules/mob/living/simple_animal/borer/borer_chemicals.dm [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+/datum/borer_chem/morphine
+	chemname = "morphine"
+	needed_influence = 50
+	influence_change = -10
+	chemuse = 100
+
+/datum/borer_chem/spacedrugs
+	chemname = "space_drugs"
+	needed_influence = 50
+	chemuse = 75
```
Comment (ShadowDeath6) [inline code/modules/mob/living/simple_animal/borer/borer_chemicals.dm:36]: Not entirely familiar with code, but if you added ephedrine to the available list, do you need to add it here, too?

Comment (oisin100) [inline code/modules/mob/living/simple_animal/borer/borer_chemicals.dm:36]: Woopsies. Forgot to add ephedrine!

**Comment by oisin100** [comment]
Alright, Borers will be spawned in random vents. Due to git being retarded.

File: code/modules/mob/living/simple_animal/borer/borer.dm [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+/mob/living/simple_animal/borer/New()
+	..()
+	name = "[pick("Primary","Secondary","Tertiary","Quaternary")] Borer ([rand(100,999)])"
+	borer_chems += /datum/borer_chem/mannitol
+	borer_chems += /datum/borer_chem/bicardine
+	borer_chems += /datum/borer_chem/kelotane
+	borer_chems += /datum/borer_chem/charcoal
+	borer_chems += /datum/borer_chem/leporazine
+	borer_chems += /datum/borer_chem/morphine
+	borer_chems += /datum/borer_chem/spacedrugs
```
Comment (X-TheDark) [inline code/modules/mob/living/simple_animal/borer/borer.dm:41]: I think this should be replaced with actual chems (as in, create a chem using new() then add it to this list), so you can operate with that directly in `borer_powers.dm`, when you create injection choice list, etc, because right now, it seems this is the only place borer_chems is used (which means it's not actually used at all).

Comment (oisin100) [inline code/modules/mob/living/simple_animal/borer/borer.dm:41]: borer_chems is used in borer_powers. They need to be datums as some chems cost different amounts of chemicals to use and also have different influence requirements. I could use if statements in borer_powers but that would just make it harder in the long run to maintain.

**Comment by oisin100** [comment]
Not exactly WIP anymore. But it could certainly do with some looking over by another coder to see if i have missed anything.

**Comment by Alek2ander** [comment]
Please squash the commits on this, otherwise the map changes add 300k pointless lines to the history

**Comment by oisin100** [comment]
> otherwise the map changes add 300k pointless lines to the history

Uhm, Theres only 700 total line additions for me....

**Comment by Alek2ander** [comment]
All of your commits get added to the history, and even if the final line change count is low, the pull can be much bigger

https://github.com/yogstation13/yogstation/pull/384.patch has 312660 lines and is 11.7 Mb in size.

**Comment by oisin100** [comment]
Yikes! Ill see what i can do then!

**Comment by Alek2ander** [comment]
Easiest way is to make a new branch from dev and merge this branch with the --squash option, then make a new PR with the new branch. Alternatively, I can merge this manually later today in the same way.

**Comment by oisin100** [comment]
Dammit. Im terrible at Git.

**Comment by oisin100** [comment]
Just gonna manually transfer my code to another branch... My Commits are all fucked now...

**Comment by oisin100** [comment]
#434
